src/hydration/hydrate.ts

```diff
@@ -0,0 +1,134 @@
+import { Float } from 'java.lang';
+import { Material } from 'org.bukkit';
+import { Levelled } from 'org.bukkit.block.data';
+import { Player } from 'org.bukkit.entity';
+import { Action } from 'org.bukkit.event.block';
+import {
+  PlayerInteractEvent,
+  PlayerItemConsumeEvent,
+} from 'org.bukkit.event.player';
+import { EquipmentSlot } from 'org.bukkit.inventory';
+
+const Hydration = {
+  MAX: 0.99,
+  MIN: 0.0,
+  HUGE: 0.4,
+  MEDIUM: 0.25,
+  SMALL: 0.05,
+};
+
+// When player drinks (bottles etc)
+registerEvent(PlayerItemConsumeEvent, (event) => {
+  const item = event.item;
+
+  const material = item?.type;
+  const player = event.player;
+  switch (material) {
+    case Material.POTION:
+      hydrate(player, Hydration.MEDIUM, material);
+      break;
+    case Material.MELON_SLICE:
+      hydrate(player, Hydration.SMALL, material);
+      break;
+  }
+});
+
+// When player drinks from block
+const drinkers = new Set<Player>();
+registerEvent(PlayerInteractEvent, (event) => {
+  if (
+    (event.action === Action.RIGHT_CLICK_BLOCK ||
+      event.action === Action.RIGHT_CLICK_AIR) &&
+    event.hand === EquipmentSlot.HAND
+  ) {
+    const player = event.player;
+    const material = event.item?.type;
+
+    if (material !== Material.AIR) return;
+    if (!drinkers.has(player)) return;
+
+    // Drinking from cauldron
+    const block = event.clickedBlock;
+    if (block?.type === Material.CAULDRON) {
+      const cauldronData = block.blockData as Levelled;
+      const waterLevel = cauldronData.level;
+      if (waterLevel) {
+        cauldronData.setLevel(waterLevel - 1);
+        block.setBlockData(cauldronData);
+      } else {
+        // Cauldron was empty
+        return;
+      }
+
+      drinkers.add(player);
+      hydrate(player, Hydration.MEDIUM, Material.CAULDRON);
+      playDrinkingSound(player);
+
+      stopDrinking(player, 1);
```

Comment:
Don't ignore promises returned by async functions. In this case, it would not cause any harm - but in general, it tends to introduce race conditions. Without await, your code may not be executing in the order you think it does.